### PR TITLE
[IOTDB-4541] Improve: cache latest snapshot info for ratis leader

### PR DESCRIPTION
**Pull request by SzyWilliam**
Ratis Leader will query latest snapshot info every time it sends a AppendEntries. We should cache the latest snapshot and return directly, instead of query statemachine and scanning disk.

### CODE REVIEW

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/SnapshotStorage.java [hunk 1]
```diff
@@ -135,6 +137,28 @@ public SnapshotInfo getLatestSnapshot() {
...
+  Snapshot cache will be updated upon:
+  1. initialize, when RaftServer first starts
+  2. reinitialize, when RaftServer resumes from pause. Leader may install a new snapshot during pause.
+  3. takeSnapshot, when RaftServer takes a new snapshot
+  4. loadSnapshot, when RaftServer is asked to load a snapshot
+   */
+  void updateSnapshotCache() {
+    snapshotCacheGuard.writeLock().lock();
+    currentSnapshot = findLatestSnapshot();
+    snapshotCacheGuard.writeLock().unlock();
```
Comment (JackieTien97) [inline consensus/src/main/java/org/apache/iotdb/consensus/ratis/SnapshotStorage.java:150]:
```suggestion
    try {
      currentSnapshot = findLatestSnapshot();
    } finally {
      snapshotCacheGuard.writeLock().unlock();
    }
```

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/SnapshotStorage.java [hunk 2]
```diff
@@ -135,6 +137,28 @@ public SnapshotInfo getLatestSnapshot() {
...
+  }
+
+  @Override
+  public SnapshotInfo getLatestSnapshot() {
+    SnapshotInfo latestSnapshot;
+    snapshotCacheGuard.readLock().lock();
+    latestSnapshot = currentSnapshot;
+    snapshotCacheGuard.readLock().unlock();
```
Comment (JackieTien97) [inline consensus/src/main/java/org/apache/iotdb/consensus/ratis/SnapshotStorage.java:158]: same as above.

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/SnapshotStorage.java [hunk 3]
```diff
@@ -135,6 +137,28 @@ public SnapshotInfo getLatestSnapshot() {
         fileInfos, snapshotTermIndex.getTerm(), snapshotTermIndex.getIndex());
   }
 
+  /*
+  Snapshot cache will be updated upon:
+  1. initialize, when RaftServer first starts
+  2. reinitialize, when RaftServer resumes from pause. Leader may install a new snapshot during pause.
+  3. takeSnapshot, when RaftServer takes a new snapshot
+  4. loadSnapshot, when RaftServer is asked to load a snapshot
```
Comment (JackieTien97) [inline consensus/src/main/java/org/apache/iotdb/consensus/ratis/SnapshotStorage.java:145]: update the java doc, only need to update cache at the last step of `takeSnapshot` and first step of `loadSnapshot`. `loadSnapshot` will be called by `initialize` and `reinitialize`.

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/SnapshotStorage.java (hunk 1 above)
Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/SnapshotStorage.java]: done

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/SnapshotStorage.java (hunk 2 above)
Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/SnapshotStorage.java]: done

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/SnapshotStorage.java (hunk 3 above)
Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/SnapshotStorage.java]: updated

**Review by JackieTien97 — APPROVED** [review]

**Pull request merged**